elasticsearch/templates/statefulset.yaml

```diff
@@ -191,6 +191,48 @@ spec:
           - name: node.{{ $role }}
             value: "{{ $enabled }}"
           {{- end }}
+        {{- if and (eq .Values.roles.master "true") (ge (int .Values.esMajorVersion) 7) }}
+        lifecycle:
+          preStop:
+            exec:
+              command:
+              - sh
+              - -c
+              - |
+                #!/usr/bin/env bash -e
+                # Exclude node from voting config to prevent https://github.com/elastic/helm-charts/issues/63
+                # reference: https://www.elastic.co/guide/en/elasticsearch/reference/7.0/modules-discovery-adding-removing-nodes.html#modules-discovery-removing-nodes
+                http () {
+                    local path="${1}"
+                    local method="${2:-GET}"
+                    if [ -n "${ELASTIC_USERNAME}" ] && [ -n "${ELASTIC_PASSWORD}" ]; then
+                      BASIC_AUTH="-u ${ELASTIC_USERNAME}:${ELASTIC_PASSWORD}"
+                    else
+                      BASIC_AUTH=''
+                    fi
+                    curl -X${method} -s -k --fail ${BASIC_AUTH} {{ .Values.protocol }}://127.0.0.1:{{ .Values.httpPort }}${path}
+                }
+
+                NODE_NAME=$(awk 'BEGIN {print ENVIRON["node.name"]}')
+                echo "Exclude the node ${NODE_NAME} from voting config"
+
+                http "/_cluster/voting_config_exclusions/${NODE_NAME}" "POST"
+
+                while true ; do
+                  echo -e "Wait for new master node to be elected"
+                  if [[ "$(http "/_cat/master")" != *"${NODE_NAME}"* ]]; then
+                    break
+                  fi
+                  sleep 1
+                done
+
+                # Node won't be actually removed from cluster,
+                # so node should be deleted from voting config exclusions
+                echo "Node ${NODE_NAME} is now being deleted from voting config exclusions"
+                http "/_cluster/voting_config_exclusions?wait_for_removal=false" "DELETE"
```

Comment:
This should also be inside the `while true` loop. If this fails then then this node won't be able to rejoin the cluster after starting up again.